Helper/ClientEventHelper.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ * @copyright   2018 Mautic Contributors. All rights reserved
+ * @author      Digital Media Solutions, LLC
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace MauticPlugin\MauticContactClientBundle\Helper;
+
+use Doctrine\ORM\EntityManager;
+use Mautic\CampaignBundle\Entity\CampaignRepository;
+use Mautic\CampaignBundle\Entity\EventRepository;
+use Mautic\CampaignBundle\Entity\LeadEventLogRepository;
+use Mautic\CoreBundle\Helper\CacheStorageHelper;
+
+/**
+ * Class ClientEventHelper.
+ */
+class ClientEventHelper
+{
+    /** @var CampaignRepository */
+    private $campaignRepo;
+
+    /** @var LeadEventLogRepository */
+    private $eventLogRepo;
+
+    /** @var EventRepository */
+    private $campaignEventRepo;
+
+    /** @var EntityManager */
+    private $em;
+
+    public function __construct(
+        CampaignRepository $campaignRepo,
+        LeadEventLogRepository $eventLogRepo,
+        EventRepository $campaignEventRepo,
+        EntityManager $em
+    ) {
+        $this->campaignRepo         = $campaignRepo;
+        $this->eventLogRepo         = $eventLogRepo;
+        $this->campaignEventRepo    = $campaignEventRepo;
+        $this->em                   = $em;
+    }
+
+    public function getAllClientEvents($clientId= null)
+    {
+        $cacheHelper = new CacheStorageHelper(
+            CacheStorageHelper::ADAPTOR_DATABASE,
+            'contactClientEventsMap',
+            $this->em->getConnection(),
+            null,
+            600
+        );
+        if (empty($contactClientEventsMap = $cacheHelper->get('contactClientEventsMap', 600))) {
```

Comment:
Recommend you set the cache to an empty array while the query is running to avoid stampeding the heavy queries.